File: backend/services/rfpLogic.js
function parseUsDate(dateStr) {
  if (!dateStr || typeof dateStr !== 'string') return null
  const s = dateStr.trim()
  if (!s || s.toLowerCase() === 'not available') return null
  const m = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/)
  if (!m) return null
  const mm = parseInt(m[1], 10)
  const dd = parseInt(m[2], 10)
  const yyyy = parseInt(m[3], 10)
  if (!mm || !dd || !yyyy) return null
  const d = new Date(yyyy, mm - 1, dd)
  if (Number.isNaN(d.getTime())) return null
  if (d.getFullYear() !== yyyy || d.getMonth() !== mm - 1 || d.getDate() !== dd)
    return null
  return d
}

function daysUntil(d, now = new Date()) {
  if (!d) return null
  const ms = d.getTime() - now.getTime()
  return Math.ceil(ms / (1000 * 60 * 60 * 24))
}

function computeDateSanity(rfp) {
  const now = new Date()
  const warnings = []
  const meta = { dates: {} }

  const fields = [
    ['submissionDeadline', 'Submission deadline'],
    ['questionsDeadline', 'Questions deadline'],
    ['bidMeetingDate', 'Bid meeting date'],
    ['bidRegistrationDate', 'Bid registration date'],
    ['projectDeadline', 'Project deadline'],
  ]

  for (const [key, label] of fields) {
    const raw = rfp?.[key]
    const parsed = parseUsDate(raw)
    if (raw && typeof raw === 'string' && raw !== 'Not available' && !parsed) {
      warnings.push(`${label} looks invalid (${raw}).`)
    }
    if (parsed) {
      const du = daysUntil(parsed, now)
      meta.dates[key] = {
        raw,
        iso: parsed.toISOString(),
        daysUntil: du,
        isPast: du !== null ? du < 0 : null,
      }
      // Only hard-warn on submission deadline being past; other dates are informational.
      if (key === 'submissionDeadline' && du !== null && du < 0) {
        warnings.push(`${label} appears past (${raw}).`)
      }
    }
  }

  return { warnings, meta }
}

function checkDisqualification(rfp) {
  const now = new Date()
  const sub = parseUsDate(rfp?.submissionDeadline)
  if (sub && sub < now) return true

  const raw = typeof rfp?.rawText === 'string' ? rfp.rawText.toLowerCase() : ''
  const isMandatoryMeeting =
    raw.includes('mandatory') &&
    (raw.includes('pre-bid') ||
      raw.includes('prebid') ||
      raw.includes('pre-proposal') ||
      raw.includes('preproposal') ||
      raw.includes('site visit') ||
      raw.includes('bid conference') ||
      raw.includes('pre proposal conference'))
  const isMandatoryRegistration =
    raw.includes('mandatory') &&
    (raw.includes('registration') ||
      raw.includes('vendor registration') ||
      raw.includes('bid registration') ||
      raw.includes('register'))

  if (isMandatoryMeeting) {
    const meeting = parseUsDate(rfp?.bidMeetingDate)
    if (meeting && meeting < now) return true
  }
  if (isMandatoryRegistration) {
    const reg = parseUsDate(rfp?.bidRegistrationDate)
    if (reg && reg < now) return true
  }

  return false
}

function computeFitScore(rfp) {
  const now = new Date()
  const reasons = []
  let score = 100

  const raw = typeof rfp?.rawText === 'string' ? rfp.rawText.toLowerCase() : ''
  const sub = parseUsDate(rfp?.submissionDeadline)
  const q = parseUsDate(rfp?.questionsDeadline)
  const meeting = parseUsDate(rfp?.bidMeetingDate)
  const reg = parseUsDate(rfp?.bidRegistrationDate)

  // Hard disqualifier
  if (sub && sub < now) {
    return {
      score: 0,
      reasons: ['Submission deadline passed.'],
      disqualified: true,
    }
  }

  const duSub = sub ? daysUntil(sub, now) : null
  if (typeof duSub === 'number') {
    if (duSub <= 7) {
      score -= 20
      reasons.push(`Due soon (${duSub} days until submission).`)
    } else if (duSub <= 14) {
      score -= 10
      reasons.push(`Moderately urgent (${duSub} days until submission).`)
    }
  }

  const duQ = q ? daysUntil(q, now) : null
  if (typeof duQ === 'number' && duQ < 0) {
    score -= 10
    reasons.push('Questions deadline appears past.')
  }

  const isMandatoryMeeting = raw.includes('mandatory') && raw.includes('pre')
  if (isMandatoryMeeting) {
    if (!meeting) {
      score -= 10
      reasons.push(
        'Mentions mandatory pre-bid meeting but no meeting date found.',
      )
    } else {
      const du = daysUntil(meeting, now)
      if (typeof du === 'number' && du < 0) {
        return {
          score: 0,
          reasons: ['Mandatory meeting appears past.'],
          disqualified: true,
        }
      }
      reasons.push('Mandatory pre-bid meeting detected.')
      score -= 5
    }
  }

  const isMandatoryRegistration =
    raw.includes('mandatory') && raw.includes('register')
  if (isMandatoryRegistration) {
    if (!reg) {
      score -= 10
      reasons.push(
        'Mentions mandatory registration but no registration date found.',
      )
    } else {
      const du = daysUntil(reg, now)
      if (typeof du === 'number' && du < 0) {
        return {
          score: 0,
          reasons: ['Mandatory registration appears past.'],
          disqualified: true,
        }
      }
      reasons.push('Mandatory registration detected.')
      score -= 5
    }
  }

  if (raw.includes('bid bond') || raw.includes('performance bond')) {
    score -= 10
    reasons.push('Bid/performance bond requirements detected.')
  }
  if (
    raw.includes('license') ||
    raw.includes('licensing') ||
    raw.includes('certification')
  ) {
    score -= 5
    reasons.push('Licensing/certification requirements detected.')
  }

  score = Math.max(0, Math.min(100, score))
  if (reasons.length === 0) reasons.push('No major risks detected.')
  return { score, reasons, disqualified: false }
}

module.exports = {
  parseUsDate,
  computeDateSanity,
  checkDisqualification,
  computeFitScore,
}

File: backend/app/services/rfp_logic.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


def parse_us_date(date_str: Any) -> datetime | None:
    if not isinstance(date_str, str):
        return None
    s = date_str.strip()
    if not s or s.lower() == "not available":
        return None

    try:
        mm, dd, yyyy = s.split("/")
        m = int(mm)
        d = int(dd)
        y = int(yyyy)
    except Exception:
        return None

    if m <= 0 or d <= 0 or y <= 0:
        return None

    try:
        dt = datetime(y, m, d, tzinfo=timezone.utc)
    except Exception:
        return None

    # sanity: ensure components match (e.g. 02/31 invalid)
    if dt.year != y or dt.month != m or dt.day != d:
        return None

    return dt


def days_until(dt: datetime | None, now: datetime | None = None) -> int | None:
    if not dt:
        return None
    now = now or datetime.now(timezone.utc)
    delta = dt - now
    return int((delta.total_seconds() + 86400 - 1) // 86400)


def compute_date_sanity(rfp: dict[str, Any]) -> dict[str, Any]:
    now = datetime.now(timezone.utc)
    warnings: list[str] = []
    meta: dict[str, Any] = {"dates": {}}

    fields = [
        ("submissionDeadline", "Submission deadline"),
        ("questionsDeadline", "Questions deadline"),
        ("bidMeetingDate", "Bid meeting date"),
        ("bidRegistrationDate", "Bid registration date"),
        ("projectDeadline", "Project deadline"),
    ]

    for key, label in fields:
        raw = rfp.get(key)
        parsed = parse_us_date(raw)
        if isinstance(raw, str) and raw and raw != "Not available" and not parsed:
            warnings.append(f"{label} looks invalid ({raw}).")
        if parsed:
            du = days_until(parsed, now)
            meta["dates"][key] = {
                "raw": raw,
                "iso": parsed.isoformat().replace("+00:00", "Z"),
                "daysUntil": du,
                "isPast": (du < 0) if du is not None else None,
            }
            if key == "submissionDeadline" and du is not None and du < 0:
                warnings.append(f"{label} appears past ({raw}).")

    return {"warnings": warnings, "meta": meta}


def check_disqualification(rfp: dict[str, Any]) -> bool:
    now = datetime.now(timezone.utc)
    sub = parse_us_date(rfp.get("submissionDeadline"))
    if sub and sub < now:
        return True

    raw = str(rfp.get("rawText") or "").lower()

    is_mandatory_meeting = "mandatory" in raw and any(
        k in raw
        for k in (
            "pre-bid",
            "prebid",
            "pre-proposal",
            "preproposal",
            "site visit",
            "bid conference",
            "pre proposal conference",
        )
    )
    is_mandatory_registration = "mandatory" in raw and any(
        k in raw for k in ("registration", "vendor registration", "bid registration", "register")
    )

    if is_mandatory_meeting:
        meeting = parse_us_date(rfp.get("bidMeetingDate"))
        if meeting and meeting < now:
            return True

    if is_mandatory_registration:
        reg = parse_us_date(rfp.get("bidRegistrationDate"))
        if reg and reg < now:
            return True

    return False


def compute_fit_score(rfp: dict[str, Any]) -> dict[str, Any]:
    now = datetime.now(timezone.utc)
    reasons: list[str] = []
    score = 100

    raw = str(rfp.get("rawText") or "").lower()
    sub = parse_us_date(rfp.get("submissionDeadline"))
    q = parse_us_date(rfp.get("questionsDeadline"))
    meeting = parse_us_date(rfp.get("bidMeetingDate"))
    reg = parse_us_date(rfp.get("bidRegistrationDate"))

    if sub and sub < now:
        return {"score": 0, "reasons": ["Submission deadline passed."], "disqualified": True}

    du_sub = days_until(sub, now) if sub else None
    if isinstance(du_sub, int):
        if du_sub <= 7:
            score -= 20
            reasons.append(f"Due soon ({du_sub} days until submission).")
        elif du_sub <= 14:
            score -= 10
            reasons.append(f"Moderately urgent ({du_sub} days until submission).")

    du_q = days_until(q, now) if q else None
    if isinstance(du_q, int) and du_q < 0:
        score -= 10
        reasons.append("Questions deadline appears past.")

    is_mandatory_meeting = "mandatory" in raw and "pre" in raw
    if is_mandatory_meeting:
        if not meeting:
            score -= 10
            reasons.append("Mentions mandatory pre-bid meeting but no meeting date found.")
        else:
            du = days_until(meeting, now)
            if isinstance(du, int) and du < 0:
                return {"score": 0, "reasons": ["Mandatory meeting appears past."], "disqualified": True}
            reasons.append("Mandatory pre-bid meeting detected.")
            score -= 5

    is_mandatory_registration = "mandatory" in raw and "register" in raw
    if is_mandatory_registration:
        if not reg:
            score -= 10
            reasons.append("Mentions mandatory registration but no registration date found.")
        else:
            du = days_until(reg, now)
            if isinstance(du, int) and du < 0:
                return {"score": 0, "reasons": ["Mandatory registration appears past."], "disqualified": True}
            reasons.append("Mandatory registration detected.")
            score -= 5

    if "bid bond" in raw or "performance bond" in raw:
        score -= 10
        reasons.append("Bid/performance bond requirements detected.")

    if any(k in raw for k in ("license", "licensing", "certification")):
        score -= 5
        reasons.append("Licensing/certification requirements detected.")

    score = max(0, min(100, score))
    if not reasons:
        reasons.append("No major risks detected.")

    return {"score": score, "reasons": reasons, "disqualified": False}
